### ISSUE TITLE
[ISSUE][CLOSED] "devtool add" commmand issue while integrating Yocto build module

### ISSUE THREAD

**Issue by leemgs** [leemgs/TAOS-CI14#139]
<a href="https://github.sec.samsung.net/geunsik-lim"><img src="https://github.sec.samsung.net/avatars/u/8170?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [geunsik-lim](https://github.sec.samsung.net/geunsik-lim)**
_Wednesday Jul 18, 2018 at 05:49 GMT_
_Originally opened as https://github.sec.samsung.net/STAR/TAOS-CI/issues/136_

----

It's strange. 
* How to evaluate:
   * build test in console (with www-data id) : okay
   * build test via apache webserver (with www-data id): failed (devtool add ..... is not executed.)

* log messages: 
   * Running Yocto build module via web (id: www-data)
```bash
########################################################################################
[MODULE] TAOS/pr-audit-build-smp: check build process for SMP distribution
[DEBUG] source /var/www/kairos_sdk/environment-setup-core2-64-smp-linux
SDK environment now set up; additionally you may now run devtool to perform development tasks.
Run devtool --help for further details.
Checking for sudo...
/usr/bin/sudo
Checking for curl...
/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/curl
Checking for devtool...
/var/www/kairos_sdk/sysroots/x86_64-smpsdk-linux/usr/bin/devtool
[DEBUG] starting TAOS/pr-audit-build-smp facility
[DEBUG] The current folder is /var/www/html/smp-hello-world/TAOS-CI/ci/repo-workers/pr-audit/201807181444560.37642407417297-2-54143e4eec68a24afc2a83603c87826f65bad4ca/smp-hello-world.
[DEBUG] Checking /var/www/kairos_sdk/workspace/recipes/smp-hello-world-54143e4eec68a24afc2a83603c87826f65bad4ca
[DEBUG] devtool reset smp-hello-world-54143e4eec68a24afc2a83603c87826f65bad4ca
Checking for devtool...
/var/www/kairos_sdk/sysroots/x86_64-smpsdk-linux/usr/bin/devtool
uid=33(www-data) gid=33(www-data) groups=33(www-data)
[DEBUG] devtool add smp-hello-world-54143e4eec68a24afc2a83603c87826f65bad4ca git@github.sec.samsung.net:geunsik-lim/smp-hello-world.git
[DEBUG] The return value of 'devtool add' command is 1.
[DEBUG] starting 'devtool build' command.
[DEBUG] devtool build smp-hello-world-54143e4eec68a24afc2a83603c87826f65bad4ca 
[DEBUG] The return value (build_result) of 'devtool build' command is 1.
BUILD_MODE != 99
[DEBUG] The return value of devtool command is 1.
[DEBUG] Oooops!!!!!! SMP build checker is failed. Return value is (1).
```
   * stdout(1)
```
Nothing.
```
   * stderr(2)
```bash
Please use a locale setting which supports utf-8.
Python can't change the filesystem locale after loading so we need a utf-8 when python starts or things won't work.
```

* Screenshot:

![image](https://github.sec.samsung.net/storage/user/8170/files/e295786e-8aa9-11e8-9001-7f6d6d9e4019)

![image](https://github.sec.samsung.net/storage/user/8170/files/248b7dae-8aaa-11e8-8e24-667eea176425)

**Comment by leemgs** [comment]
<a href="https://github.sec.samsung.net/git-bot-sec"><img src="https://github.sec.samsung.net/avatars/u/63289?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [git-bot-sec](https://github.sec.samsung.net/git-bot-sec)**
_Wednesday Jul 18, 2018 at 05:49 GMT_

----

:octocat: **cibot**: Thank you for posting issue #136. The person in charge will reply soon.

**Comment by leemgs** [comment]
<a href="https://github.sec.samsung.net/geunsik-lim"><img src="https://github.sec.samsung.net/avatars/u/8170?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [geunsik-lim](https://github.sec.samsung.net/geunsik-lim)**
_Wednesday Jul 18, 2018 at 07:26 GMT_

----

> [DEBUG] devtool add smp-hello-world-54143e4eec68a24afc2a83603c87826f65bad4ca git@github.sec.samsung.net:geunsik-lim/smp-hello-world.git
[DEBUG] The return value of 'devtool add' command is 1.

We have to find a reason why "**devtool add**" command cannot run normally in case of `www-data` id of Apache webserver.

**Comment by leemgs** [comment]
<a href="https://github.sec.samsung.net/geunsik-lim"><img src="https://github.sec.samsung.net/avatars/u/8170?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [geunsik-lim](https://github.sec.samsung.net/geunsik-lim)**
_Wednesday Jul 18, 2018 at 07:48 GMT_

----

* Analysis: 

```bash
ubuntu16.04$ id -a
uid=33(www-data) gid=33(www-data) groups=33(www-data)
ubuntu16.04$ env
AS=x86_64-smp-linux-as
LC_PAPER=ko_KR.UTF-8
LDFLAGS=-Wl,-O1
AR=x86_64-smp-linux-ar
OECORE_TARGET_SYSROOT=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
TARGET_PREFIX=x86_64-smp-linux-
LC_ADDRESS=ko_KR.UTF-8
XDG_SESSION_ID=90
M4=m4
LC_MONETARY=ko_KR.UTF-8
COMP_WORDBREAKS=
"'><;|&(:
TERM=xterm
SHELL=/bin/bash
NM=x86_64-smp-linux-nm
OECORE_DISTRO_VERSION=1.0a
CPPFLAGS=
SSH_CLIENT=10.113.62.204 13925 22
LC_NUMERIC=ko_KR.UTF-8
OLDPWD=/var/www/kairos_sdk/workspace
GIT_SSL_CAINFO=/var/www/kairos_sdk/buildtools/sysroots/x86_64-smpsdk-linux/etc/ssl/certs/ca-certificates.crt
SSH_TTY=/dev/pts/1
OECORE_SDK_VERSION=1.0
GDB=x86_64-smp-linux-gdb
http_proxy=http://10.112.1.184:8080/
USER=www-data
LC_TELEPHONE=ko_KR.UTF-8
PKG_CONFIG_SYSROOT_DIR=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
OE_SKIP_SDK_CHECK=1
CXXFLAGS=
CONFIG_SITE=/var/www/kairos_sdk/site-config-core2-64-smp-linux
CPP=x86_64-smp-linux-gcc -E  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse --sysroot=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
MAIL=/var/mail/www-data
PATH=/var/www/kairos_sdk/sysroots/x86_64-smpsdk-linux/usr/bin:/var/www/kairos_sdk/buildtools/sysroots/x86_64-smpsdk-linux/usr/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/sbin:/var/www/kairos_sdk/tmp/sysroots/x86_64/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/sbin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/../x86_64-smpsdk-linux/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux-uclibc:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux-musl:/var/www/kairos_sdk/sysroots/x86_64-smpsdk-linux/usr/bin:/var/www/kairos_sdk/buildtools/sysroots/x86_64-smpsdk-linux/usr/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/sbin:/var/www/kairos_sdk/tmp/sysroots/x86_64/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/sbin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/../x86_64-smpsdk-linux/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux-uclibc:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux-musl:/var/www/kairos_sdk/sysroots/x86_64-smpsdk-linux/usr/bin:/var/www/kairos_sdk/buildtools/sysroots/x86_64-smpsdk-linux/usr/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/sbin:/var/www/kairos_sdk/tmp/sysroots/x86_64/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/sbin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/../x86_64-smpsdk-linux/bin:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux-uclibc:/var/www/kairos_sdk/tmp/sysroots/x86_64/usr/bin/x86_64-smp-linux-musl:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
QT_QPA_PLATFORMTHEME=appmenu-qt5
LD=x86_64-smp-linux-ld   --sysroot=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
LC_IDENTIFICATION=ko_KR.UTF-8
PWD=/var/www/html/smp-hello-world/TAOS-CI/ci/repo-workers/pr-audit/201807181132450.33490705490112-2-54143e4eec68a24afc2a83603c87826f65bad4ca/smp-hello-world
STRIP=x86_64-smp-linux-strip
DEPLOY_DIR_IMAGE=/var/www/kairos_sdk/tmp/deploy/images/genericx86-64
LANG=en_US.UTF-8
LC_MEASUREMENT=ko_KR.UTF-8
OECORE_NATIVE_SYSROOT=/var/www/kairos_sdk/tmp/sysroots/x86_64
CONFIGURE_FLAGS=--target=x86_64-smp-linux --host=x86_64-smp-linux --build=x86_64-linux --with-libtool-sysroot=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
https_proxy=https://10.112.1.184:8080/
KCFLAGS=--sysroot=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
CXX=x86_64-smp-linux-g++  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse --sysroot=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
OBJCOPY=x86_64-smp-linux-objcopy
SHLVL=1
HOME=/var/www
CFLAGS= -O2 -pipe
no_proxy=localhost,127.0.0.1,165.213.149.200,10.113.136.32,10.113.136.201,github.sec.samsung.net,www.samsung.net,kr1.samsung.net
LOGNAME=www-data
SSH_CONNECTION=10.113.62.204 13925 10.113.138.43 22
OECORE_ACLOCAL_OPTS=-I /var/www/kairos_sdk/tmp/sysroots/x86_64/usr/share/aclocal
PKG_CONFIG_PATH=/var/www/kairos_sdk/tmp/sysroots/genericx86-64/usr/lib/pkgconfig:/var/www/kairos_sdk/tmp/sysroots/genericx86-64/usr/share/pkgconfig
ARCH=x86
RANLIB=x86_64-smp-linux-ranlib
CROSS_COMPILE=x86_64-smp-linux-
CC=x86_64-smp-linux-gcc  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse --sysroot=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
XDG_RUNTIME_DIR=/run/user/33
DISPLAY=localhost:11.0
OBJDUMP=x86_64-smp-linux-objdump
LC_TIME=ko_KR.UTF-8
SDKTARGETSYSROOT=/var/www/kairos_sdk/tmp/sysroots/genericx86-64
LC_NAME=ko_KR.UTF-8
_=/usr/bin/env

```

* The difference between console command and Apache web command
   
[env_console.txt](https://github.sec.samsung.net/STAR/TAOS-CI/files/2362/env_console.txt)
[env_web.txt](https://github.sec.samsung.net/STAR/TAOS-CI/files/2363/env_web.txt)

**Comment by leemgs** [comment]
<a href="https://github.sec.samsung.net/geunsik-lim"><img src="https://github.sec.samsung.net/avatars/u/8170?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [geunsik-lim](https://github.sec.samsung.net/geunsik-lim)**
_Wednesday Jul 18, 2018 at 08:57 GMT_

----

Here is recipe that is written by @sewon-oh 

```bash
 export LC_ALL=en_US.UTF-8
 export LANG=en_US.UTF-8
 export LANGUAGE=en_US.UTF-8
```

**Comment by leemgs** [comment]
<a href="https://github.sec.samsung.net/geunsik-lim"><img src="https://github.sec.samsung.net/avatars/u/8170?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [geunsik-lim](https://github.sec.samsung.net/geunsik-lim)**
_Wednesday Jul 18, 2018 at 09:03 GMT_

----

https://github.sec.samsung.net/STAR/TAOS-CI/pull/125 fixed this issue. 

Note that you have to declare language set to avoid the execution error of "devtool add/build" command because  Python can not change the filesystem locale after loading so we need a utf-8 when python starts or things won't work. Use a locale setting which supports utf-8.
*  See https://github.com/openembedded/openembedded-core/blob/master/scripts/devtool#L212
* See https://github.com/openembedded/bitbake/blob/master/bin/bitbake#L38


According to Python community, print() and write() (and maybe other of such I/O functions) are relying on sys.getfilesystemencoding().
For more details, visit the below webpages.
* https://bugs.python.org/issue19846
* https://vstinner.github.io/pep-383.html

**Comment by leemgs** [comment]
<a href="https://github.sec.samsung.net/geunsik-lim"><img src="https://github.sec.samsung.net/avatars/u/8170?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [geunsik-lim](https://github.sec.samsung.net/geunsik-lim)**
_Thursday Jul 19, 2018 at 04:10 GMT_

----

Fixed by PR #125.
